ENH: Support for statistical analysis of whole containers

pastewka opened the pull request:
- All topography/line scan classes now have a `unit` attribute and a `unit` keyword argument for the constructor
- The `scale` pipeline function has support for converting between units
- Container have a pipeline function `scale_dependent_statistical_property` that runs a whole container statistical analysis

pastewka commented:
Hi @mcrot - I need your help here. For some reason, `ScaledTopography` can no longer be pickled. When unpickling, properties are missing. This is super weird, I have never seen this before. Do you have any ideas what may be going on?

pastewka commented:
I found the problem but the error was very unintuitive

pastewka commented:
Hi @mcrot - this is now ready for your review

mcrot commented on SurfaceTopography/Container/ScaleDependentStatistics.py:
```diff
@@ -0,0 +1,57 @@
+from .SurfaceContainer import SurfaceContainer
+
+
+def scale_dependent_statistical_property(container, func, n, distance, unit):
```
Here a docstring would be helpful to understand what the function does and what the arguments mean.

pastewka replied:
I updated the doc strings

mcrot commented on SurfaceTopography/HeightContainer.py:
```diff
@@ -63,7 +63,8 @@ class Error(Exception):
 
-    def __init__(self, info={}, communicator=MPI.COMM_WORLD):
+    def __init__(self, unit=None, info={}, communicator=MPI.COMM_WORLD):
+        self._unit = unit
```
Two Ideas: 

- We could show a deprecation warning here if `info` still has a unit attribute. Would uncover if some of our tests still rely on `info`
- We should raise an error if `info` and `unit` and it's not equal to the given `unit`. This would uncover if some tests weren't correctly initialized.

Old code would not be affected, since both `unit` and `info['unit']` need to be given for a warning/error.

pastewka replied:
This is a good idea. I implemented this. It will raise an error if the parameter is passed in the info dictionary.

mcrot commented on SurfaceTopography/IO/Reader.py:
```diff
@@ -221,13 +224,20 @@ def area_per_pt(self):
         The area per point is returned as the product over the pixel size
         tuple. If `pixel_size` returns None, than also `area_per_pt` returns
-        None.
+        None.s
```
Maybe a typo?

pastewka replied:
Yes a typo. Fixed this.

mcrot commented on SurfaceTopography/NonuniformLineScan.py:
```diff
@@ -132,41 +164,91 @@ def squeeze(self):
+            Factor to scale lateral positions (`physical_sizes`, etc.) with.
+            If None, scale factor will be autodetermined from the given unit.
+            (Default: None)
+        unit : str
```
What if only one scale factor is given and the other not, with or without a given `unit`?

pastewka replied:
The default is (or rather should be) unity for the scale factors.

pastewka replied:
I just saw that this is actually explained in the unit docstring. Is this sufficiently clear?

mcrot replied:
I also saw it, but it leaves some cases open. 
There are 2³=8 cases for providing arguments which we need to handle here (3x None or not None)
in the code, either by throwing an exception or taking the right values. Please see the attached drawing:
![cases](https://user-images.githubusercontent.com/9113541/123064467-9509a500-d40e-11eb-89a9-058b760deb55.jpg)

It seems to me that if not both scale factors are given and no unit is given, an exception should be raised (first three cases). If both factors are given, the unit is not needed (4th case).
If only the unit is given, the scale factors can be determined from that unit (5th case).

My question refers to the green question marks for the last three cases.
Which factors are taken if not both scale factors are given and unit is given?

pastewka replied:
I think the best solution is to move the unit conversion to a separate pipeline function `to_unit` and simply patch through unit in the `scale` pipeline function. Then it is clear what's going on.

mcrot commented on SurfaceTopography/NonuniformLineScan.py:
```diff
@@ -132,41 +164,91 @@ def squeeze(self):
+        if self._position_scale_factor is None:
+            if self.parent_topography.unit is None:
+                if self.unit is not None:
+                    raise ValueError('You are trying to rescale a unitless topography to a topography with a unit')
```
The user just requests the scale factor for positions, why is there an exception saying that the user tries to rescale?

If this is a check whether the configuration of the topography is valid, this should be done in the constructor,
because otherwise the topography can be used although it is invalid.

pastewka replied:
What is meant by this error is the following: You are running `t.scale(unit='um')`, i.e. you want to rescale to 'um' but `t` does not have a unit. This is only possible if `t` has a unit. I made the error clearer.

pastewka replied:
I've also added the error to the constructor

pastewka replied:
(Although the parent topography may change. So constructor may not be able to decide this.)

pastewka replied:
(This is in theory, in the current implementation of the code this cannot happen.)

mcrot replied:
> What is meant by this error is the following: You are running `t.scale(unit='um')`, i.e. you want to rescale to 'um' but `t` does not have a unit. This is only possible if `t` has a unit. I made the error clearer.

But the exception is thrown in the property `position_scale_factor`, not in the `scale` method?
However, it is now also in the constructor and that should be sufficient.

Maybe we could do this kind of check for all height containers automatically.

mcrot commented on SurfaceTopography/Uniform/ScaleDependentStatistics.py:
```diff
@@ -0,0 +1,38 @@
+from ..HeightContainer import UniformTopographyInterface
+
+
+def scale_dependent_statistical_property(topography, func, n=1, scale_factors=None, distance=None):
```
A docstring would be helpful here, I think.

pastewka replied:
I added this docstring

mcrot commented on SurfaceTopography/Uniform/VariableBandwidth.py:
```diff
@@ -32,7 +32,30 @@
 
 
-def checkerboard_detrend(topography, subdivisions):
+def checkerboard_detrend_1D(topography, region_index, order=1, return_plane=False):
```
For other functions we used the postfixes `from_area` and `from_profile`. Would this apply here, too?

pastewka replied:
This part of the code has been updated. There is now a `checkerboard_detrend_profile` and `checkerboard_detrend_area` and a `variable_bandwidth_from_area` and `variable_bandwidth_from_profile`. They all should have docstrings.

mcrot commented on SurfaceTopography/Uniform/common.py:
```diff
@@ -134,13 +134,12 @@ def _trim_nonperiodic(arr, scale_factor, op):
         if r == 0:
             r = None
-        trimmed_slice += [slice(L, r)]
+        trimmed_slice += [slice(int(L), int(r))]
```
`int(None)` doesn't work:
```python
In [1]: int(None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-af2123a46eb2> in <module>
----> 1 int(None)

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```
This can happen here?!

pastewka replied:
Good catch! I think it can but there is no test for it... I just fixed the code

mcrot commented on SurfaceTopography/Uniform/common.py:
```diff
@@ -150,10 +149,15 @@ def derivative(topography, n, scale_factor=1, operator=None, periodic=None,
     n : int
         Order of the derivative.
-    scale_factor : int or list of ints, optional
+    scale_factor : int or list of ints or list of tuples of ints, optional
```
Is it sufficiently clear, what giving a "list of tuples of ints"  means here in terms of a scale factor? 
It's a rather complex argument value.

pastewka replied:
I extended the docstring to explain what the tuples do

mcrot commented on test/IO/test_nc.py:
```diff
@@ -42,15 +42,15 @@ def test_save_and_load(comm):
                             fft.nb_subdomain_grid_pts,
                             communicator=comm)
     assert t.info['unit'] == 'µm'
+    assert dt.info['unit'] == 'µm'
```
Here maybe also
```python
assert t.unit == 'µm'
```

pastewka replied:
Added this!

mcrot commented on test/IO/test_opdx.py:
```diff
@@ -133,7 +133,7 @@ def test_read_header(self):
         self.assertEqual(channel_1.info['Height_value'], 35.85522403809594)
         self.assertEqual(channel_1.info['z_scale'], 78.592625)
 
-        assert channel_1.info['unit'] == 'µm'  # see GH 281
```
Shouldn't we leave this in the test, until we really remove `unit` from `info`? 

Which GH 281 is meant here - SurfaceTopography is around 100?

pastewka replied:
I guess this is from the old PyCo issue tracker: https://github.com/pastewka/PyCo/issues/100 - I've readded the test

mcrot commented on test/test_derivatives.py:
```diff
@@ -97,7 +97,7 @@ def test_scale_factor():
 
     d1 = topography.derivative(1, scale_factor=1)
-    d2 = topography.derivative(1, scale_factor=2)
+    d2 = topography.derivative(1, scale_factor=np.uint32(2))
```
Can the other type of scale factors also be tested here?
From the docstring: "int or list of ints or list of tuples of ints"

pastewka replied:
I added this test

mcrot commented on test/test_topography.py:
```diff
@@ -96,9 +96,9 @@ def test_attribute_error(self):
 
         with self.assertRaises(AttributeError):
-            t.scale_factor
+            t.heights_scale_factor
```
Is here a typo? `t.height_scale_factor` without `s`?

pastewka replied:
Good catch, fixed this

mcrot reviewed (changes requested):
Please find my comments in the code.

Additionally, I suggest to extend the file `test_api.py` by the new interface components:
- `unit`
- `scale_dependent_statistical_property`
- `checkerboard_detrend_1D` and `2D` if neeed (or with `from_area`.. if better)

pastewka commented:
Hi @mcrot - I think I have addressed all your comments. Can you have a look at the current PR again? Note that I cleaned up the CI configuration to make as much use of Ubuntu's packages as possible.

mcrot commented on SurfaceTopography/Container/ScaleDependentStatistics.py:
```diff
@@ -0,0 +1,57 @@
+                    retvals[e] += [s]
+                else:
+                    retvals[e] = [s]
+    return [np.mean(retvals[d], axis=0) for d in distance]
```
This can still be a problem, can't it? Isn't it possible to pass distance array which raise `KeyError` in the last line?

pastewka replied:
Yes you are right... sorry overlooked this, will fix...

pastewka replied:
I just fixed this. I am now returning None if no data exists at a certain scale.

pastewka commented:
Hi @mcrot - I changed the behavior to issue a deprecation warning when a unit is passed in the info dictionary. Also, you should get a deprecation warning now if you access the unit key of the info dictionary.

mcrot commented on SurfaceTopography/NonuniformLineScan.py:
```diff
@@ -188,7 +188,7 @@ def height_scale_factor(self):
     # For backwards compatibility
     @property
-    @deprecated(version='0.94.0', reason='Use heights_scale_factor')
+    @deprecated(version='0.94.0', alternative='heights_scale_factor')
```
The property in the code is called `height_scale_factor`, without "s".
There are more places affected by this (search for "heights_scale_factor").

mcrot commented on SurfaceTopography/NonuniformLineScan.py:
```diff
@@ -259,7 +259,7 @@ def height_scale_factor(self):
     # For backwards compatibility
     @property
-    @deprecated(version='0.94.0', reason='Use heights_scale_factor')
+    @deprecated(version='0.95.0', alternative='heights_scale_factor')
```
Also here.

mcrot commented on SurfaceTopography/Support/Deprecation.py:
```diff
@@ -0,0 +1,64 @@
+from functools import wraps
+
+
+def deprecated(version=None, alternative=None):
```
Just an idea: Like the original "deprecated" one could add a sentence at the end of the doc string
about the deprecation, the same string as the warning when using it.

Like this our Sphinx documentation would show what is deprecated.

pastewka replied:
I am not sure I know how to do that. I tried adding something to `func.__doc__`, but that does not show up in the help string.

mcrot replied:
Hm... The package "deprecated" seems to do it in a similar way: https://github.com/tantale/deprecated/blob/09c31517c5f864ab07819eb4ec2235671ad43854/deprecated/sphinx.py#L120

I've tried it like that and it seems to work, I'll push it to this branch.

mcrot replied:
Let me know what you think - either we remove it or simply close this conversation issue.

mcrot replied:
Just another comment for starting a review (just understood how this works here).

mcrot commented on SurfaceTopography/UniformLineScanAndTopography.py:
```diff
@@ -572,7 +572,7 @@ def height_scale_factor(self):
     # For backwards compatibility
     @property
-    @deprecated(version='0.94.0', reason='Use heights_scale_factor')
+    @deprecated(version='0.95.0', alternative='heights_scale_factor')
```
Also here without "s".

mcrot commented on SurfaceTopography/UniformLineScanAndTopography.py:
```diff
@@ -494,7 +494,7 @@ def height_scale_factor(self):
     # For backwards compatibility
     @property
-    @deprecated(version='0.94.0', reason='Use heights_scale_factor')
+    @deprecated(version='0.95.0', alternative='heights_scale_factor')
```
Also here without "s".

mcrot reviewed (changes requested):
Looks good, except the wrong property name in the deprecation messages.
Please see the comments.

mcrot commented:
Oh, there are more changes that I haven't reviewed yet, please wait..

pastewka commented:
I've fixed heights_scale_factor to height_scale_factor

mcrot commented on SurfaceTopography/Support/Deprecation.py:
```diff
@@ -0,0 +1,64 @@
+    def deprecated_decorator(func):
+        @wraps(func)
+        def deprecated_func(*args, **kwargs):
+            if version is None:
```
Here "then" and "else" parts are interchanged. I'll send a patch together with the doc string thing.

mcrot replied:
Just another comment for starting a review (just understood how this works here).

mcrot reviewed (approved):
Please review my changes to the deprecation message and the 
last unresolved conversation. If this is okay for you, then this PR should be good.

mcrot commented on SurfaceTopography/Support/Deprecation.py:
```diff
@@ -0,0 +1,64 @@
+                warn_str = f'Function/property `{func.__name__}` was deprecated in SurfaceTopography {version}.'
+            else:
+                warn_str = f'Function/property `{func.__name__}` is deprecated.'
+            if alternative is None:
```
Fixed. Okay like this?

pastewka replied:
Looks good

mcrot reviewed (approved):
Please review the open conversations and my fixes for the deprecation decorator.

pastewka commented on SurfaceTopography/Uniform/Filtering.py:
```diff
@@ -201,7 +201,7 @@ def heights(self):
 
             qx = np.arange(0, nx, dtype=np.float64).reshape(-1, 1)
-            qx = np.where(qx <= nx // 2, qx / sx, (nx - qx) / sx)
+            qx = np.where(qx <= nx // 2, qx / sx,  - (nx - qx) / sx)
```
But we could switch this to nx-qx

AntoineSIMTEK replied:
qx -nx you mean

pastewka replied:
I switched this

pastewka commented on test/test_api.py:
```diff
@@ -36,7 +36,10 @@
+    "scale",
     "squeeze",
+    "scale_dependent_statistical_property",
+    "to_unit",
```
Good point, added that

pastewka commented:
Hi @mcrot - I am happy with your changes to the deprecation module and have addressed the other issues you raised
